src/terrain/services/sharing.clj

```diff
@@ -172,12 +172,25 @@
       path-list
       (str "unable to send unshare error notification for " unsharee))))
 
+(defn- get-user-from-subject
+  [subject]
+  (condp = (:source_id subject)
+    "ldap"  (:id subject)
+    "g:gsa" (str "@grouper-" (:id subject))
+    nil))
+
+(defn- translate-user-for-irods
+  [share-unshare]
+  (let [provided-user (:user share-unshare)
+        provided-subject (:subject share-unshare)]
+    (or provided-user (get-user-from-subject provided-subject))))
+
 (defn- share-with-user
   "Forwards share requests to data-info from the user and list of paths and permissions in the given
    share map, sending any success notifications to the users involved, and any error notifications
    to the current user."
   [share]
-  (let [user (:user share)
+  (let [user (translate-user-for-irods share)
```

Comment:
I figure that if for some reason something passes neither of user/subject, this'll just spit out a nil and that'll throw an error when it tries to pass it down the chain. I could also be more explicit about validation if desired though.